src/mpid/ch4/generic/mpidig.h

```diff
@@ -29,6 +29,7 @@ typedef int (*MPIDIG_am_target_msg_cb)
 typedef struct MPIDIG_global_t {
     MPIDIG_am_target_msg_cb target_msg_cbs[MPIDI_AM_HANDLERS_MAX];
     MPIDIG_am_origin_cb origin_cbs[MPIDI_AM_HANDLERS_MAX];
+    MPI_Aint nm_eager_max;
```

Comment:
In my opinion, the active message layer should not differentiate between netmod or shmmod other than at the site of making the transport call. Deciding whether to do eager protocol or rendezvous protocol sits above the protocols. So I am concerned having `nm_eager_max` that forcing the decision inside resulting in nested protocols. I understand the current master is essentially doing nested protocols, but this is something I am aiming to change.

Ideally, we'll have eager send protocol -- started by `MPIDIG_am_send_eager`, synchronous protocol -- started by `MPIDIG_am_ssend`, and rendezvous protocol -- started by `MPIDIG_am_send_long` (the names can be improved). The protocol decisions are made in `MPIDIG_mpi_send` (if we do generic fallback) or in `OFI_mpi_send/UCX_mpi_send/POSIX_mpi_send` (if we do selective fallback/forward inside netmod/shmmod).  The current code in `MPIDIG_isend_impl` (essentially `MPIDIG_mpi_send`) is in fact doing protocol specific stuff such as filling the active message header, which is protocol specific and should only be touched inside each protocol. Therefore, your implementation is still making protocol choices inside a protocol.

Reply:
> Ideally, we'll have eager send protocol -- started by `MPIDIG_am_send_eager`, synchronous protocol -- started by `MPIDIG_am_ssend`, and rendezvous protocol -- started by `MPIDIG_am_send_long` (the names can be improved). The protocol decisions are made in `MPIDIG_mpi_send` (if we do generic fallback) or in `OFI_mpi_send/UCX_mpi_send/POSIX_mpi_send` (if we do selective fallback/forward inside netmod/shmmod). 

Yes, I think nm/shm making the protocol decisions is best. What I'm missing in my head is how to inform the generic layer of the message type. Say ch4/ucx wants to use`MPIDIG_am_send_long` to do a `MPI_PUT` operation, how is that expressed in the API?

Reply:
I am not sure I understand your scenario. `MPIDIG_am_send_long` is a rendezvous protocol for pt2pt send/recv. `MPI_PUT` is a one-sided operation, it can't directly use the pt2pt rendezvous protocol as that protocol need an explicit recv.

`MPI_PUT` won't need rendezvous as the recv-buffer is always ready, right?

I think what you mean is to implement MPI_PUT in a rendezvous-similar fashion, with an automatic buffer matching? If that is desired, it should be implemented as a separate protocol, e.g. `MPIDIG_am_put_long`, then `MPI_PUT` in ch4/ucx should call `MPIDIG_am_put_long` directly.

If we desire to share code between `MPIDIG_am_send_long` and `MPIDIG_am_put_long`, we can use an inline header file, e.g. `mpidig_msg_long.h` for common functionality. But the protocol code, include struct type for am_hdr, callback functions, handler_ids should be totally independent.

Reply:
> how to inform the generic layer of the message type.

The message type is determined by the protocol. Different message type (that require different handling) should always be implemented as different protocols.

Reply:
> 
> `MPI_PUT` won't need rendezvous as the recv-buffer is always ready, right?

To avoid extra copies there needs to be coordination with the target. In my mind this is still rendezvous. Rendezvous != tag matching.

> 
> I think what you mean is to implement MPI_PUT in a rendezvous-similar fashion, with an automatic buffer matching? If that is desired, it should be implemented as a separate protocol, e.g. `MPIDIG_am_put_long`, then `MPI_PUT` in ch4/ucx should call `MPIDIG_am_put_long` directly.

Yes, this is what I am getting at. It seems we need additional AM APIs to support different message types (PUT, GET, Atomic, other?) efficiently.

Reply:
Currently we have separate protocols for put, get, cswap, acc, get_acc. All the window synchronizations are in one protocol.

Reply:
This is sounding a little like the UCP API, which would be a good thing IMO.

Reply:
The separate rma protocols are currently in the master, in case you are not aware :)

Reply:
Can you provide a link? I do not see any RMA active message APIs in ch4 at this time. I'm not sure if you are being snarky.

Reply:
Here we have all the handler ids:
https://github.com/pmodels/mpich/blob/26703ade1e963340c92c8e7846c6e0b1bd835e06/src/mpid/ch4/src/ch4_types.h#L56-L90
And each protocol has all the separate callbacks in `ch4r_rma_{origin,target}_callbacks.{h,c}`.

Reply:
They have each different "am_hdr" structure too, but again in a separate header file with no opaqueness.

Reply:
Sure, those are the message types which inform the transports which handler to call. At the moment, they all layer on top of the am_send_hdr/am_isend "protocol" defined by the transports. Are there changes we need to make to the nm/shm active message APIs to support an expanded generic API?

Reply:
`am_send_hdr/am_isend` are not protocols, they are the transport. They protocols are implemented (currently) in e.g. MPIDIG_mpi_put (or maybe MPIDIG_put_impl) and their corresponding callbacks.

> Are there changes we need to make to the nm/shm active message APIs to support an expanded generic API?

I would like to simplify them to make the protocol layer cleaner; but no, the current API supports separate protocols. I am not sure I understand what do you mean by "expanded generic API".

Reply:
> `am_send_hdr/am_isend` are not protocols, they are the transport. The protocols are implemented (currently) in e.g. MPIDIG_mpi_put (or maybe MPIDIG_put_impl) and their corresponding callbacks.

What I am trying to say is `MPIDI_NM_am_isend` as implemented today has its own protocol. `ch4/ofi` adds an additional hdr and handlers, even doing its own rndv to move message contents.

> 
> > Are there changes we need to make to the nm/shm active message APIs to support an expanded generic API?
> 
> I would like to simplify them to make the protocol layer cleaner; but no, the current API supports separate protocols. I am not sure I understand what do you mean by "expanded generic API".
 
You listed new generic APIs like `MPIDIG_am_put_long`. Are these all calling `MPIDI_NM_am_isend` underneath or do we need new transport functionality to support them?

Reply:
> You listed new generic APIs like `MPIDIG_am_put_long`. Are these all calling `MPIDI_NM_am_isend` underneath or do we need new transport functionality to support them?

Yes, `MPIDIG_am_xxx` are protocol entry functions, they call lower-layer transport such as `MPIDI_NM_am_isend`. 

`MPIDIG_am_put_long` (or in current code, it would be simply `MPIDIG_am_put`) is just a wrapper for code currently in `MPIDIG_put`. I suggest the wrapper to make the protocols have better isolations.